internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -736,3 +778,73 @@ func getComposerResourcesNames(cds []ComposedResource) []string {
 	}
 	return names
 }
+
+type compositionResultMeta struct {
+	numWarningEvents   int
+	conditionTypesSeen map[xpv1.ConditionType]struct{}
+}
+
+func handleCommonCompositionResult(ctx context.Context, r *Reconciler, res CompositionResult, xr *composite.Unstructured, log logging.Logger) compositionResultMeta {
```

Comment:
Perhaps make this a method on `Reconciler`? It looks like it uses the reconciler's client, event handler, and logger.

Reply:
I'm also wondering if it might be better to just "inline" this into the two places it's called. A little repetitive, but maybe easier to follow. I often prefer a little repetition over a little abstraction.

Reply:
I turned this into a method as suggested by the first comment. It feels pretty long to inline IMO, and abstracting it prevents a good amount of nesting.

If you feel strongly about inlining it, LMK and I'll change it.